### DISCUSSION TITLE
Accessor API for report returned by `@check`

### DISCUSSION THREAD

**Discussion by Seelengrab** [Seelengrab/Supposition.jl#57] [Design & Ideas]
Currently, the report returned by `@check` is an opaque object with lots of internals. This can be quite annoying when there is some programmatic manipulation or inspection required, such as 

   * when a `NaN` with a non-standard bitpattern causes a property to fail (it's all printed as `NaN`, so that hides the failure)
   * when a user wants to build on top of `@check` without going through the entire `AbstractTestset` machinery for wrapping.
   * when hypothetical statistics about the generation & shrinking process are interesting of themselves, for example for a visualization 

For cases like these, it would be good to have an official API for retrieving that information from a report.

**Comment by aplavin** [comment]


**Reply by Seelengrab** [reply]
I still wonder what `ispass` would do for you that regular invocation of `@check` doesn't already cover :thinking: Like I mentioned on zulip, `@check` already integrates into the Test stdlib, so doing something like `@test ispass(@check ...)` seems quite odd to me..

If there's some other framework Supposition.jl could be integrated into, I'd be happy to help with their respective API :)

**Reply by aplavin** [reply]
My personal preference is always to have a fully-featured base interface without any involved framework integration, to be used in isolation. Then, framework integrations are built on top – they are clearly useful, but best if kept optional.
Looking from this PoV, interface in the spirit of `ispass(check(f, gens...))` makes a lot of sense – as the most basic usage interface, not relying on being inside testsets.
(or `ispass(@check ...)` if you really don't want a function-based interface)

**Reply by jariji** [reply]
> interface in the spirit of `ispass(check(f, gens...))`

This function-based interface appeals to me as well. 

Even if I use `@check` much of the time, it's nice to have access to a simpler function-only `check(f, gens...)` form for when I want to compose checks with other tools.

**Reply by Seelengrab** [reply]
So, `ispass(::SuppositionReport) -> Bool` is very unlikely to ever happen. I don't think this is a good API - I'll add something like `counterexample(::SuppositionReport) -> Option{T}` (name pending) instead, with `T` being the type of what would be passed/splatted into the property. There are a few reasons for this:

 * There is no need for `ispass` internally - whenever I need to work with the result at that stage, I *have* to handle all cases. `ispass` doesn't help with that; if anything, I'd need a proper sum-type. Barring that, dispatch on other internal types is perfectly sufficient.
 * There is very little you can do with `ispass` if you have a `SuppositionReport` - `true` is mostly uninteresting, and `false` would immediately require getting the actual result for further processing. At that point, it's more convenient to use `@something` to unpack the `Option{T}` and continue, or `break`/`return` out of whatever you're doing (or otherwise define `ispass` yourself, if you really want that - bear in mind that this might cause more (re)generation to happen).

Of course, this still doesn't make `@test counterexample(...)` a useful thing to do - but at the same time, it's not supposed to be in the first place!

As for a functional version of `@check` - I still don't see what this enables a user to do, and I think t would rather have an overall negative impact. For example, for type stability `@check` creates a unique function to actually generate the input passed to the property. This is not possible with a functional interface, because all invocations of `check` would need to share the function that generates the input to the property, which is inherently type unstable once there are more than 4 invocations of `check` due to the return type heuristics giving up. When invoking a property thousands of times or more in one `@check`, this adds noticeable overhead for small properties (which the vast majority should be, precisely to enable running them with thousands of distinct inputs!).

I understand that in principle a functional interface _can_ be preferable, but there are lots of tradeoffs to be made that need to be considered. It's not as simple as personal preference, so if you really want that, I'd need an actual usecase for this functionality, or a PR implementing it (which does not guarantee that I'll merge it - in the end, I have to maintain it after all).

**Reply by aplavin** [reply]
As always, thanks for your detailed response, and for laying out your vision of the user-facing interface! It's totally fine to disagree, and this is your package that doesn't have to cater to everyone. We – potential users – just share our opinions here, nothing more, and don't attempt to force you choose a design you are not happy with.

---
Re `ispass`: of course the naming was just an example :) Any public way to obtain the status of a `check` would be generally useful, and it's less important whether it's called `ispass(r)` or `isnothing(counterexample(r))`.

Re `check` vs `@check`: tbh, I don't see the problem here. What is wrong with a simple implementation of
```julia
julia> check(f, gens...) = @check f(gens...)

julia> check(i -> i < 100, Data.Integers{Int8}())
Found counterexample
  Context: f

  Arguments:
      arg_1::Int8 = 100
```
?

**Reply by Seelengrab** [reply]
> of course the naming was just an example :) Any public way to obtain the status of a check would be generally useful, and it's less important whether it's called ispass(r) or isnothing(counterexample(r)). 

Gotcha - thank you for clarifying your usecase! I want to reiterate though that this is not just a naming thing, this is user-facing API. There needs to be more thought put into this than "it feels right", which is why the returned type not being `Bool` is important.

> Re check vs @check: tbh, I don't see the problem here. What is wrong with a simple implementation of

I've so far taken your requests for a functional interface as a replacement for `@check` since you mention to use functional elements as the building blocks for an API, that's all. If you think about it as an additional API, that may be more feasible, but I still don't see how this is an improvement UX-wise. You can't meaningfully chain `@check` with itself (like you could with e.g. `filter` or `map`), and I just don't see what meaningful operation a functional API enables a user to do that is not possible with `@check`.

For example, a common usecase for a functional API is using `do`-notation, but that just makes the code look like a worse version of `@check`, which would require at least one additional method to allow passing a `CheckConfig`. Also, the functional API has the possibility for getting a mismatch in the number of declared arguments vs. the number of passed arguments, leading to confusing `MethodError`s. This can't happen as easily in `@check` because the function definition itself is right there, coupling the passed arguments to the variables - it's correct by design, and nothing needs to be checked (of course, if you use an existing property you can still get a `MethodError`, but it's not the norm to use the API like that!). In contrast, the functional API would _require_ reflection to achieve the same, which is a comparatively brittle solution.

This is the kind of UX tradeoff a functional API has, and so far I'm not convinced that a functional API would bring enough benefit to balance this out. As I mentioned above, if you really want it, please state your usecase (i.e., _what_ you want to do, not _how_ you want to do it) and we can discuss further from there.

**Reply by aplavin** [reply]
> I want to reiterate though that this is not just a naming thing, this is user-facing API. 

My original point was just that it's useful to have a public/documented way to extract the status of a `check`. Further details don't matter _that_ much. Even just guaranteeing and documenting the current `Supposition.results(x).ispass` would be helpful.

**Reply by aplavin** [reply]
As for function vs macro, I guess it's just a matter of personal perspective. My view is that the "default" choice is/should always be function-based. And macro should only supplement it, e.g. by giving a more convenient syntax. It's very rare that a macro user-facing interface is _required_, and `@check` doesn't appear to be one of those cases – my `check()` definition above seems to work just as well as `@check`. Functions are more transparent and are not susceptible to parsing/hygiene/scoping issues that macros may have.

>  I just don't see what meaningful operation a functional API enables a user to do that is not possible with @check

> I'm not convinced that a functional API would bring enough benefit to balance this out.

I tend to think the other way around: if a functional API is good enough, then macros should be optional.

> if you really want it, please state your usecase 

I don't really have a special usecase... Tried Supposition once in https://github.com/JuliaAPlavin/Uncertain.jl/blob/master/test/maths.jl to check consistency: setup in https://github.com/JuliaAPlavin/Uncertain.jl/blob/master/test/maths.jl#L12-L34 and usage in https://github.com/JuliaAPlavin/Uncertain.jl/blob/master/test/maths.jl#L42-L51.

**Reply by Seelengrab** [reply]
> My original point was just that it's useful to have a public/documented way to extract the status of a check

Right, that was a misunderstanding on my part then - you seemed very insistent on the Bool-returning API.

> Further details don't matter that much.

API design matters a great deal! Choose wrong and you're stuck with it until a breaking change. If you're lucky it's just unergonomic to use; if you're unlucky, the wrong choice undermines the guarantees you're trying to provide.

> As for function vs macro, I guess it's just a matter of personal perspective. My view is that the "default" choice is/should always be function-based. 

Yes, and I agree with that. That's why PropCheck.jl had a functional interface! While using it for more extensive testing though, it quickly became quite cumbersome and turned out to be more trouble than it was worth.

> my `check()` definition above seems to work just as well as `@check`

Or put differently - `@check` works just as well as a functional API, and hence an additional functional API needs to fulfill a higher bar to be included at this point :wink: 

> Tried Supposition once in

These links 404 for me :thinking: Cool to hear that it helped you though!

---

Either way, the main point of contention around `ispass` is resolved I believe. I'll ping this thread again once it's on `main` :)

**Reply by aplavin** [reply]
> you seemed very insistent on the Bool-returning API

Yes, I want to get a Bool, but don't care how the API looks exactly: `ispass(r)`, `isnothing(counterexpample(r))`, `Supposition.results(r).ispass`, etc.

**Reply by aplavin** [reply]
> These links 404 for me 🤔

Sorry, fixed the links! Would be curious to hear your thoughts on my usage of Supposition there – does it look reasonable?

> Or put differently - @check works just as well as a functional API, and hence an additional functional API needs to fulfill a higher bar to be included at this point 😉

It's just a common recommendation in Julia to avoid defining macros unless really needed, as functions are more transparent in their behavior.
Of course, there's nothing functions can do that macros cannot. Any function in Julia could have been a macro instead, but why would one define e.g. `@filter()` when `filter()` is just as good? :)

**Reply by Seelengrab** [reply]
> It's just a common recommendation in Julia to avoid defining macros unless really needed, as functions are more transparent in their behavior.

I have been around long enough in the Julia community to know what I'm doing with macros, thank you. This is far from my first rodeo.

> Sorry, fixed the links!

Thanks! I see now - do you want that API as a workaround for https://github.com/julia-vscode/TestItemRunner.jl/issues/99? I really think this is an issue on the end of that package, and not in Supposition.jl. Would be great to have someone from the VSCode team look into this, from my POV this should just work out of the box. After all, it works with the standard `@testset` :shrug: 

> does it look reasonable?

The property looks fine for the most part, yes. A few notes though:

 * Don't `try`/`catch` around the function you actually want to check, Supposition.jl can already handle errors for you & shrink the inputs correctly, preserving the error down to the location it was thrown from. As you've written it, you're hiding errors that happen in that code completely, which you might want to report upstream (and use as promotional material for your own, provided it can handle those inputs).
 * Prefer `filter!` over manually calling `reject!` in your property, and similarly prefer `map(U.Value, Data.Floats{Float32}(), Data.Floats{Float32}())` over `@composed` (not because it's a functional interface, but because it has to generate a struct to get type stability. It's much more sensitive to compiler changes than `map` because of that).
 * If I'm reading this correctly, your access of internals & `@test ispass` construct should result in any result to be reported twice to the containing `@testset`. For example:
![image](https://github.com/user-attachments/assets/caa5c5d4-1190-4a8f-8fad-5510ed4ad371)
 This only contains one (always passing) `@check`, but the result is reported twice due to the `@test` - `@check` is already reporting to the `@testset`! Similarly, when an error is thrown you get confusing double reporting of a single testcase as both erroring as well as failing, which doesn't really make sense:
![image](https://github.com/user-attachments/assets/bfb52def-58ee-458b-9563-5483fd1b80d2)

I wonder why you set `db=false` everywhere? You can wrap your usage in `@with DEFAULT_CONFIG => ...` to disable it for larger blocks of code at once. 

Finally, using Supposition.jl only for comparison and not for investigating more interesting properties of your functions seems like a bit of a waste to me :) Surely there are some cool invariants your operations should uphold, independent of other libraries?

**Reply by Seelengrab** [reply]
`main` now contains API for retrieving a found counterexample (if any) through `counterexample`, as well as an accessor API for various collected statistics.
